packages/custom-elements/src/CustomElementRegistry.js

```diff
@@ -261,11 +308,20 @@ export default class CustomElementRegistry {
     const deferred = new Deferred();
     this._whenDefinedDeferred.set(localName, deferred);
 
-    const definition = this._localNameToDefinition.get(localName);
-    // Resolve immediately only if the given local name has a definition *and*
-    // the full document walk to upgrade elements with that local name has
-    // already happened.
-    if (definition && !this._elementsWithPendingDefinitions.has(localName)) {
+    // Resolve immediately if the given local name has a regular or lazy
+    // definition *and* the full document walk to upgrade elements with that
+    // local name has already happened.
+    //
+    // The behavior of the returned promise differs between the lazy and the
+    // non-lazy cases if the definition fails. Normally, the definition would
+    // fail synchronously and no pending promises would resolve. However, if
+    // the definition is lazy but has not yet been reified, the promise is
```

Comment:
Likewise, below how can this localName be in `_locanNameToConstructorGetter` here?

Reply:
The name could be in `_localNameToConstructorGetter` but not represent a valid definition if the user defines it with `polyfillDefineLazy` but nothing has triggered that getter yet:

```javascript
// No <some-element>s are in the main document.
customElements.polyfillDefineLazy('some-element', () => { throw new Error(); });
customElements.whenDefined('some-element').then(() => console.log('This will run.'));
```

Later, when something creates a `<some-element>`, the getter is called and the definition fails, even though the `whenDefined` promise already resolved.

The behavior of promises from `whenDefined` for pending lazy definitions is totally new though, so we could also pick something different. WDYT about this behavior vs. others? One other option would be to wait until it's actually reified to resolve them, but I think we had a quick offline discussion where we ended up worried that it would prevent using `whenDefined` to wait until a definition exists before creating the first instance. Another option might be to make a new `whenDefined`-like thing with special behavior for lazy definitions.

Reply:
Sorry, misread this. Think it's ok.